src/test/java/seedu/address/logic/commands/DeleteAttendanceCommandTest.java

```diff
@@ -0,0 +1,169 @@
+
+        assertCommandFailure(deleteAttendanceCommand, model, Messages.MESSAGE_INVALID_MODULE_CLASS_DISPLAYED_INDEX);
+    }
+
+    @Test
+    public void execute_invalidLessonIndex_failure() {
+        Index moduleClassIndex = INDEX_FIRST_ITEM;
+        ModuleClass moduleClass = model.getFilteredModuleClassList().get(moduleClassIndex.getZeroBased());
+        Index lessonIndex = Index.fromOneBased(moduleClass.getLessons().size() + 1);
+        Index studentIndex = INDEX_FIRST_ITEM;
+        Week targetWeek = VALID_WEEK_1;
+
+        DeleteAttendanceCommand deleteAttendanceCommand =
+                new DeleteAttendanceCommand(moduleClassIndex, lessonIndex, studentIndex, targetWeek);
+
+        assertCommandFailure(deleteAttendanceCommand, model, Messages.MESSAGE_INVALID_LESSON_DISPLAYED_INDEX);
+    }
+
+    @Test
+    public void execute_invalidStudentIndex_failure() {
+        Index moduleClassIndex = INDEX_FIRST_ITEM;
+        Index lessonIndex = INDEX_FIRST_ITEM;
+        Index studentIndex = Index.fromOneBased(model.getFilteredStudentList().size() + 1);
+        Week targetWeek = VALID_WEEK_1;
+
+        DeleteAttendanceCommand deleteAttendanceCommand =
+                new DeleteAttendanceCommand(moduleClassIndex, lessonIndex, studentIndex, targetWeek);
+
+        assertCommandFailure(deleteAttendanceCommand, model, Messages.MESSAGE_INVALID_STUDENT_DISPLAYED_INDEX);
+    }
+
+    @Test
+    public void execute_invalidWeek_failure() {
+        Index moduleClassIndex = INDEX_FIRST_ITEM;
+        Index lessonIndex = INDEX_FIRST_ITEM;
+        Index studentIndex = INDEX_FIRST_ITEM;
+        ModuleClass moduleClass = model.getFilteredModuleClassList().get(moduleClassIndex.getZeroBased());
+        Lesson lesson = moduleClass.getLessons().get(lessonIndex.getZeroBased());
+        Week targetWeek =
+                new Week(Index.fromOneBased(lesson.getAttendanceRecordList().getAttendanceRecordList().size() + 1));
+
+        DeleteAttendanceCommand deleteAttendanceCommand =
+                new DeleteAttendanceCommand(moduleClassIndex, lessonIndex, studentIndex, targetWeek);
+
+        assertCommandFailure(deleteAttendanceCommand, model, Messages.MESSAGE_INVALID_WEEK);
+    }
+
+    /**
+     * Returns a new {@code ModuleClass} based on the given {@code moduleClass} but with the specified
+     * {@code targetLesson} replaced.
+     */
+    private static ModuleClass manualReplaceLessonToModuleClass(ModuleClass moduleClass,
+                                                                Lesson targetLesson, Lesson modifiedLesson) {
+        List<Lesson> lessons = new ArrayList<>(moduleClass.getLessons());
+
+        for (Lesson lesson : lessons) {
+            if (lesson.isSameLesson(targetLesson)) {
+                lessons.set(lessons.indexOf(lesson), modifiedLesson);
+            }
+        }
+
+        return new ModuleClass(moduleClass.getName(), moduleClass.getStudentUuids(), lessons);
+    }
```

Comment:
I think this appeared in `AddAttendanceCommandTest` previously. Instead of duplicating it, perhaps we can make it `public` in the other class for others to use. It can also be moved to another more suitable location, especially as it seems like these helper methods are becoming more common.

Reply:
To do this in separate PR

Reply:
Abstracted to ```ModuleClassUtil```